pygen/pygen_src/riscv_asm_program_gen.py

```diff
@@ -704,9 +745,66 @@ def gen_test_file(self, test_name):
         file.close()
         logging.info("%0s is generated", test_name)
 
-    def gen_signature_handshake(self, instr, signature_type, core_status = "INITIALIZED",
-                                test_result = "TEST_FAIL", csr = "MSCRATCH", addr_label = ""):
-        pass
+    def gen_signature_handshake(self, instr, signature_type,
+                                core_status = core_status_t.INITIALIZED,
+                                test_result = test_result_t.TEST_FAIL,
+                                csr = privileged_reg_t.MSCRATCH,
+                                addr_label = ""):
+        if cfg.require_signature_addr:
+            string = []
+            string = ("li x{}, {}".format(cfg.gpr[1].value, hex(cfg.signature_addr)))
+            instr.extends(string)
+            # A single data word is written to the signature address.
+            # Bits [7:0] contain the signature_type of CORE_STATUS, and the upper
+            # XLEN-8 bits contain the core_status_t data.
+            if signature_type == signature_type_t.CORE_STATUS:
+                string.extend(("li x{}, {}".format(cfg.gpr[0].value, hex(core_status.value)),
+                               "slli x{}, x{}, 8".format(cfg.gpr[0].value, cfg.gpr[0].value),
+                               "addi x{}, x{}, {}".format(cfg.gpr[0].value, cfg.gpr[0].value,
+                                                          hex(signature_type.value)),
+                               "sw x{}, 0(x{})".format(cfg.gpr[0].value, cfg.gpr[1].value)))
+                instr.extend(string)
```

Comment:
Is there a reason here for not extending the "instr" directly? I mean why would we need to extend the string first, then extend the instr?

Reply:
thanks for pointing this.
 it's fixed now!